### ISSUE TITLE
Ability to deploy proxy using create2

### ISSUE THREAD

**Issue by Remscar** [OpenZeppelin/openzeppelin-upgrades#358]
It would be helpful if it were possible to define a salt when calling `deployProxy` which would indicate to use CREATE2 and deploy out a proxy at a deterministic address.

Happy to elaborate more, but for instance if I wanted to deploy the same upgradable proxy on two different chains, and make it so they have the same address, it wouldn't be possible when using this library.

**Comment by frangio** [comment]
Can you elaborate how this would work? As far as I can tell we would need to deploy a create2 factory contract, which would also need to be on the same address on every network, so it's kind of a circular problem.

**Comment by frangio** [comment]
Yeah the hard thing is getting the factory in the same address on every network. As far as I know this is only possible to do reliably using ["Nick's method"](https://weka.medium.com/how-to-send-ether-to-11-440-people-187e332566b7).

The ability to deploy proxies on the same address on every network sounds good, but unfortunately we don't have enough bandwidth right now to make it a priority.

**Comment by habdelra** [comment]
Gnosis safe factories already do this, have a single contract that can exist at the same address in every network. And in fact they even have a “replay” script that anyone can use to deploy the factory in a network where it doesn’t exist yet and it will deploy to the same address. My understanding is that the resulting address for `create2` is that it is a hash of the bye code for the contract and user provided salt. So if that is the case does it really matter who deploys the factory? Could we provide a “replay” like gnosis does for their factories to let anyone deploy the factory to any network? Here is the gnosis replay https://github.com/gnosis/safe-contract-deployment-replay

**Comment by Amxx** [comment]
@habdelra, an EOA (wallet with a private key) cannot use `create2`. `create2` power can only be used by a smart contract. This is why you need a factory.

You first need to use a script similar to what gnosis does to create the factory at the right address ... and then (and only then) you can use create2 to instanciate "child-contracts".

If you want a generic factory, that is available on most networks, and can be used to deploy anything (simple contract or complex factories) you can use [this one](https://etherscan.io/address/0xfAC000a12dA42B871c0AaD5F25391aAe62958Db1).

All the details necessary to deploy it on any new chain, including hardhat/ganache chains, can be found [here](https://github.com/iExecBlockchainComputing/iexec-solidity/blob/master/deployment/factory.json)

Last shameless plug: [my devcon talk that discuss the need for such factory](https://www.youtube.com/watch?v=8w4zRtkfKT0)

**Comment by sebandres** [comment]
In case it helps someone, or for someone with extra time to tidy this up, I've butchered a bit the hardhat plugin and core package to allow for the use of a create2 contract with the uups proxy.

This is the Diff: [fork diff](https://github.com/OpenZeppelin/openzeppelin-upgrades/compare/master...sebandres:feature/create2-proxy-deploy?expand=1)

You must deploy the create2 contract first, then use it to deploy any other upgradable proxies like so:

```
// Attach to existing proxy
var proxy = new ethers.ContractFactory(factoryAbi, factoryBytecode, owner);
var deployedProxy = await proxy.attach('0x...');

// Token to be deployed behind a UUPS proxy
const MyToken = await ethers.getContractFactory("MyToken");
const myToken = await upgrades.deployProxy(MyToken, ["Token", "TKN"], { kind: 'uups', deployFactory: deployedProxy, deployFactorySalt: 'test' });
await myToken.deployed();

console.log("MyToken deployed to:", myToken.address);
```

This has the caveat of needing a generic create2 contract deployed to all networks at the same address which could be achieved with a fresh account and deploying the contract as the first transaction to ensure all nonce matches generating the same address. Once you have this factory you can deploy any contracts with pre-computed addresses.

This is the create2 contract I've used with it: [solidity-create2-deployer](https://github.com/thegostep/solidity-create2-deployer/blob/master/src/utils.ts#L6)

**Comment by tab00** [comment]
> If you want a generic factory, that is available on most networks, and can be used to deploy anything (simple contract or complex factories) you can use [this one](https://etherscan.io/address/0xfAC000a12dA42B871c0AaD5F25391aAe62958Db1).

There is also Axelar's [Constant Address Deployer](https://docs.axelar.dev/dev/solidity-utilities#constant-address-deployer) that can be used to deploy a new smart contract using CREATE2. But how can we use it to deploy an upgradeable contract?

**Comment by ethermachine** [comment]
> If you want a generic factory, that is available on most networks, and can be used to deploy anything (simple contract or complex factories) you can use [this one](https://etherscan.io/address/0xfAC000a12dA42B871c0AaD5F25391aAe62958Db1).

I am checking how to deploy using create2 on a selfdestructed contract but I see that unless I send the exact same bytecode it deploys to a new address. It's confusing because I read that only the salt and the factory address are used to generate it. Can you please clarify? thanks

**Comment by waynehoover** [comment]
The ability to deploy proxies via create2 is the last remaining thing keeping me from using the upgrades plugin to deploy, is this feature planned?

**Comment by frangio** [comment]
Not planned at the moment.

When we implement this we will need to use a standard create2 factory that is deployed on multiple chains. If someone has a favorite factory please share the link here.

**Comment by tab00** [comment]
Have you checked Axelar as mentioned above? They specialize in cross-chain.

`CREATE3` could also be worth having a look into: https://github.com/ZeframLou/create3-factory

It describes the difference between `CREATE2` and `CREATE3`:
> One could use a `CREATE2` factory that deterministically deploys contracts to an address that's unrelated to the deployer's nonce, but the address is still related to the hash of the contract's creation code. This means if you wanted to use different constructor parameters on different chains, the deployed contracts will have different addresses.
> 
> A `CREATE3` factory offers the best solution: the address of the deployed contract is determined by only the deployer address and the salt. This makes it far easier to deploy contracts to multiple chains at the same addresses.
>

**Comment by Oba-One** [comment]
Any updates on this functionality?
